source/common/network/utility.cc

```diff
@@ -88,24 +88,63 @@ Api::IoCallUint64Result receiveMessage(uint64_t max_rx_datagram_size, Buffer::In
   return result;
 }
 
-} // namespace
-
-Address::InstanceConstSharedPtr Utility::parseInternetAddressNoThrow(const std::string& ip_address,
-                                                                     uint16_t port, bool v6only) {
+StatusOr<sockaddr_in> parseV4Address(const std::string& ip_address, uint16_t port) {
   sockaddr_in sa4;
   memset(&sa4, 0, sizeof(sa4));
-  if (inet_pton(AF_INET, ip_address.c_str(), &sa4.sin_addr) == 1) {
-    sa4.sin_family = AF_INET;
-    sa4.sin_port = htons(port);
-    return instanceOrNull(Address::InstanceFactory::createInstancePtr<Address::Ipv4Instance>(&sa4));
+  if (inet_pton(AF_INET, ip_address.c_str(), &sa4.sin_addr) != 1) {
+    return absl::FailedPreconditionError("failed parsing ipv4");
   }
+  sa4.sin_family = AF_INET;
+  sa4.sin_port = htons(port);
+  return sa4;
+}
+
+StatusOr<sockaddr_in6> parseV6Address(const std::string& ip_address, uint16_t port) {
   sockaddr_in6 sa6;
   memset(&sa6, 0, sizeof(sa6));
-  if (inet_pton(AF_INET6, ip_address.c_str(), &sa6.sin6_addr) == 1) {
+  const auto scope_pos = ip_address.rfind('%');
+  if (scope_pos == std::string::npos) {
+    // Parse IPv6 with no scope.
+    if (inet_pton(AF_INET6, ip_address.c_str(), &sa6.sin6_addr) != 1) {
+      return absl::FailedPreconditionError("failed parsing ipv6 with no scope");
+    }
     sa6.sin6_family = AF_INET6;
-    sa6.sin6_port = htons(port);
+  } else {
```

Comment:
I think having a single path for parsing IP address is a win for simplicity. If you are concerned about possible disruption of this PR, I'd recommend adding a runtime flag to guard this.

Reply:
resolving this one and we can continue the discussion on the comment above.

Reply:
Given the different behavior of windows, at this point I would rather decouple this fix from adopting the `getaddrinfo` for the parsing while there is a resolution to the discussion that I started on slack.